File: Samples/DecodeJpeg2000WithMemoryLimit/VB.NET/DecodeJpeg2000WithMemoryLimit.vb
Imports System
Imports System.Diagnostics
Imports BitMiracle.Jpeg2k
Imports BitMiracle.LibTiff.Classic

Module DecodeJpeg2000WithMemoryLimit
    Sub Main(ByVal args As String())
        ' The library won't work without a license. You can get free time limited license
        ' at https://bitmiracle.com/jpeg2000/
        LicenseManager.SetTrialLicense("contact support@bitmiracle.com for a license")

        Const OutputFileName As String = "output.tiff"
        Using image = New J2kImage("Sample data/map.jp2")

            Using outputTiff As Tiff = Tiff.Open(OutputFileName, "w")
                Dim tileWidth As Integer = roundUp(Math.Min(2000, image.Width / 5), 16)
                Dim tileHeight As Integer = roundUp(Math.Min(2000, image.Height / 5), 16)

                Dim tileStripSize As Integer = 0
                Dim tileData As Byte() = Nothing
                Dim options = New J2kDecodingOptions()

                Dim y As Integer = 0
                While y < image.Height

                    Dim x As Integer = 0
                    While x < image.Width
                        Const TileFileName As String = "tile.tiff"
                        Dim imageData = image.DecodeArea(x, y, tileWidth, tileHeight, options)
                        imageData.Save(TileFileName, J2kOutputFormat.Tiff)

                        Using tileTiff = Tiff.Open(TileFileName, "r")

                            If x = 0 AndAlso y = 0 Then
                                Dim samplesPerPixel As Integer = tileTiff.GetField(TiffTag.SAMPLESPERPIXEL)(0).ToInt()
                                Dim bitsPerSample As Integer = tileTiff.GetField(TiffTag.BITSPERSAMPLE)(0).ToInt()
                                Dim photometric As Integer = tileTiff.GetField(TiffTag.PHOTOMETRIC)(0).ToInt()
                                outputTiff.SetField(TiffTag.IMAGEWIDTH, image.Width)
                                outputTiff.SetField(TiffTag.IMAGELENGTH, image.Height)
                                outputTiff.SetField(TiffTag.BITSPERSAMPLE, bitsPerSample)
                                outputTiff.SetField(TiffTag.SAMPLESPERPIXEL, samplesPerPixel)
                                outputTiff.SetField(TiffTag.ORIENTATION, Orientation.TOPLEFT)
                                outputTiff.SetField(TiffTag.PLANARCONFIG, PlanarConfig.CONTIG)
                                outputTiff.SetField(TiffTag.PHOTOMETRIC, photometric)
                                outputTiff.SetField(TiffTag.TILEWIDTH, tileWidth)
                                outputTiff.SetField(TiffTag.TILELENGTH, tileHeight)
                                tileStripSize = tileTiff.StripSize()
                                tileData = New Byte(tileTiff.NumberOfStrips() * tileStripSize - 1) {}
                            End If

                            Dim offset As Integer = 0
                            Dim stripCount As Integer = tileTiff.NumberOfStrips()

                            For i As Integer = 0 To stripCount - 1
                                Dim bytesRead As Integer = tileTiff.ReadRawStrip(i, tileData, offset, tileData.Length - offset)
                                offset += bytesRead
                                offset += (tileStripSize - bytesRead)
                            Next

                            outputTiff.WriteTile(tileData, x, y, 0, 0)
                        End Using

                        Console.WriteLine($"Processed tile {{{x}, {y}, {imageData.Width}, {imageData.Height}}}")

                        x += tileWidth
                    End While

                    y += tileHeight
                End While
            End Using
        End Using

        Process.Start(OutputFileName)
    End Sub

    Private Function roundUp(ByVal numToRound As Integer, ByVal multiple As Integer) As Integer
        If multiple = 0 Then Return numToRound

        Dim remainder As Integer = numToRound Mod multiple
        If remainder = 0 Then Return numToRound

        Return numToRound + multiple - remainder
    End Function
End Module

File: Samples/DecodeJpeg2000WithMemoryLimit/C#/DecodeJpeg2000WithMemoryLimit.cs
using System;
using System.Diagnostics;
using BitMiracle.Jpeg2k;
using BitMiracle.LibTiff.Classic;

namespace DecodeJpeg2000WithMemoryLimit
{
    class DecodeJpeg2000WithMemoryLimit
    {
        static void Main(string[] args)
        {
            // The library won't work without a license. You can get free time limited license
            // at https://bitmiracle.com/jpeg2000/
            LicenseManager.SetTrialLicense("contact support@bitmiracle.com for a license");

            const string OutputFileName = "output.tiff";
            using (var image = new J2kImage(@"Sample data/map.jp2"))
            {
                using (Tiff outputTiff = Tiff.Open(OutputFileName, "w"))
                {
                    // Tile size must be a multiple of 16.
                    int tileWidth = roundUp(Math.Min(2000, image.Width / 5), 16);
                    int tileHeight = roundUp(Math.Min(2000, image.Height / 5), 16);

                    int tileStripSize = 0;
                    byte[] tileData = null;
                    var options = new J2kDecodingOptions();
                    for (int y = 0; y < image.Height; y += tileHeight)
                    {
                        for (int x = 0; x < image.Width; x += tileWidth)
                        {
                            const string TileFileName = "tile.tiff";
                            var imageData = image.DecodeArea(x, y, tileWidth, tileHeight, options);
                            imageData.Save(TileFileName, J2kOutputFormat.Tiff);
                            using (var tileTiff = Tiff.Open(TileFileName, "r"))
                            {
                                if (x == 0 && y == 0)
                                {
                                    int samplesPerPixel = tileTiff.GetField(TiffTag.SAMPLESPERPIXEL)[0].ToInt();
                                    int bitsPerSample = tileTiff.GetField(TiffTag.BITSPERSAMPLE)[0].ToInt();
                                    int photometric = tileTiff.GetField(TiffTag.PHOTOMETRIC)[0].ToInt();

                                    outputTiff.SetField(TiffTag.IMAGEWIDTH, image.Width);
                                    outputTiff.SetField(TiffTag.IMAGELENGTH, image.Height);
                                    outputTiff.SetField(TiffTag.BITSPERSAMPLE, bitsPerSample);
                                    outputTiff.SetField(TiffTag.SAMPLESPERPIXEL, samplesPerPixel);
                                    outputTiff.SetField(TiffTag.ORIENTATION, Orientation.TOPLEFT);
                                    outputTiff.SetField(TiffTag.PLANARCONFIG, PlanarConfig.CONTIG);
                                    outputTiff.SetField(TiffTag.PHOTOMETRIC, photometric);
                                    outputTiff.SetField(TiffTag.TILEWIDTH, tileWidth);
                                    outputTiff.SetField(TiffTag.TILELENGTH, tileHeight);

                                    tileStripSize = tileTiff.StripSize();
                                    tileData = new byte[tileTiff.NumberOfStrips() * tileStripSize];
                                }

                                int offset = 0;
                                int stripCount = tileTiff.NumberOfStrips();
                                for (int i = 0; i < stripCount; ++i)
                                {
                                    int bytesRead = tileTiff.ReadRawStrip(i, tileData, offset, tileData.Length - offset);
                                    offset += bytesRead;

                                    // pad the rightmost column of tiles if necessary
                                    offset += (tileStripSize - bytesRead);
                                }

                                outputTiff.WriteTile(tileData, x, y, 0, 0);
                            }

                            Console.WriteLine($"Processed tile {{{x}, {y}, {imageData.Width}, {imageData.Height}}}");
                        }
                    }
                }
            }

            Process.Start(OutputFileName);
        }

        private static int roundUp(int numToRound, int multiple)
        {
            if (multiple == 0)
                return numToRound;

            int remainder = numToRound % multiple;
            if (remainder == 0)
                return numToRound;

            return numToRound + multiple - remainder;
        }
    }
}
